### PR TITLE
feat: asio as interface library, option to enable/disable deprecated …

### LINKED ISSUE #16: CPM
Avoid including CPM (download) if already available via `CPM_INITIALIZED`.
This allows to have a single CPM "instance" (e.g. project CPM) without the need to download/configure an additional CPM (avoiding also possible warning)

### PR DESCRIPTION
**Pull request by carlocorradini**
Part of PR #17
Fix #16 

asio as interface library, option to enable/disable deprecated asio functions, default asio version as asio-1-31-0 and avoid re-including cpm if already initiliazed

### CODE REVIEW

**Comment by carlocorradini** [comment]
@OlivierLDff `asio` is now an interface library and therefore the target `asio` does not exists anymore. Build only examples in CI?

**Comment by OlivierLDff** [comment]
Again can we split this PR, 
- the CPM_INITIALIZED thing can fit in its own PR for fast merging. 
- The new option about deprecated asio stuff can also be in a separate PR
- Updating default asio is a separate PR
- Updating PackageProject and using the lartian upstream is a PR
- We can have a general PR with renovation about the license year (we can even remove it), and all the cosmetic stuff that don't change anything.

Then we can discuss the interface vs static library. In general I think we don't want to change how asio is created as a static library, with source pre-compiled. Then we can add a CMake option to use ASIO as a pure interface. Maybe I'm missing a point, but tell me why using an interface would always be better?

**Comment by carlocorradini** [comment]
1) PR #29
2) Waiting other PR to be merged
3) PR #30 
4) PR #31 
5) What is missing other than finishing CMakeLists.txt?
   We have the GitHub workflow that automatically create a PR to update the license year on 1st January

_Asio_ is an header only library and therefore should be an **INTERFACE** library similarly to [nlohmann/json](https://github.com/nlohmann/json/blob/b36f4c477c40356a0ae1204b567cca3c2a57d201/CMakeLists.txt#L103).
Moreover almost every snippet (e.g. from CPM's [asio-standalone](https://github.com/cpm-cmake/CPM.cmake/wiki/More-Snippets#asio-standalone)) or example on how to use Asio (standalone) with CMake defines it as an **INTERFACE** library.

**Comment by OlivierLDff** [comment]
I understand the fact you need/want INTERFACE library, but at the moment this project is building asio with [ASIO_SEPARATE_COMPILATION](https://think-async.com/Asio/asio-1.22.0/doc/asio/using.html#:~:text=rather%20than%20chrono.-,ASIO_SEPARATE_COMPILATION,-Uses%20separately%20compiled).

I'm ok to add support for interface library, but the default way this project behave shouldn't change.

I suggest adding an `option` `ASIO_SEPARATE_COMPILATION` that is default to ON and then deal with the logic to support both. Like that user can choose which variants he wants.

Most example show the header only inclusion because it's easier to configure. The main goal of this repo was to take care of the "complex" so it's easy to add a compiled asio version to your project.

So if you go with the option way, please also update CI to test both cases

**Comment by carlocorradini** [comment]
@OlivierLDff What do you think if we add a boolean option (e.g. `ASIO_SEPARATE_COMPILATION`):
- If `false`, build `asio` as an `INTERFACE` library.
- If `true`, build the library as a `STATIC` library (if `BUILD_SHARED_LIBS` is not defined or false). Otherwise build the library as a `SHARED` library defining also `ASIO_DYN_LINK`.

Le me know what you think.
Nevertheless, I think the default behavior should be `INTERFACE` library since `asio` is known to be a header only library. If someone want/know that it can be built as a STATIC/SHARED must opt-in.

**Comment by OlivierLDff** [comment]
Create a first PR to introduce ASIO_SEPARATE_COMPILATION. I don't want to change the default behavior of the project. Default values are part of the "public API" of the repo. Changing a default would be breaking change and we definitely don't want that over the argument that example of asio usage are using the library as a header only library.
Another PR can follow with ASIO_DYN_LINK, same static library as default should be maintained and a new `ASIO_DYN_LINK` can be introduced with default value `BUILD_SHARED_LIBS`

File: CMakeLists.txt [hunk 1]
```diff
@@ -67,84 +74,93 @@ if(ASIO_CPM_FETCHCONTENT_COMPAT OR NOT ASIO_USE_CPM)
...
-  target_compile_features(asio PUBLIC cxx_std_11)
+  set_target_properties(
+    asio
+    PROPERTIES CXX_STANDARD 11
+               CXX_STANDARD_REQUIRED ON
+               CXX_EXTENSIONS OFF
+  )
   target_compile_definitions(
-    asio PUBLIC -DASIO_STANDALONE -DASIO_SEPARATE_COMPILATION -DASIO_NO_DEPRECATED
+    asio INTERFACE ASIO_STANDALONE $<$<BOOL:${ASIO_NO_DEPRECATED}>:ASIO_NO_DEPRECATED>
```
Comment (ClausKlein) [inline CMakeLists.txt:128]: if you want to install it as header only library, you have to install the ipp files too!

File: CMakeLists.txt [hunk 2]
```diff
@@ -67,84 +74,93 @@ if(ASIO_CPM_FETCHCONTENT_COMPAT OR NOT ASIO_USE_CPM)
...
+# Install
+#
 if(ASIO_ENABLE_INSTALL)
-  include(cmake/get_cpm.cmake)
-  CPMAddPackage("gh:OlivierLDff/PackageProject.cmake@1.9.0")
+  if(NOT CPM_INITIALIZED)
+    include(cmake/get_cpm.cmake)
+  endif()
+  CPMAddPackage("gh:TheLartians/PackageProject.cmake@1.12.0")
   packageProject(
```
Comment (ClausKlein) [inline CMakeLists.txt:164]: the header only installation are independent of architecture and should not installed to <prefix>/lib

**Comment by carlocorradini** [comment]
Please see PR https://github.com/OlivierLDff/asio.cmake/pull/35

File: CMakeLists.txt (hunk 1 above)
Comment (carlocorradini) [reply CMakeLists.txt:128]: Please see PR https://github.com/OlivierLDff/asio.cmake/pull/35
This has been already fixed in that PR

File: CMakeLists.txt (hunk 2 above)
Comment (carlocorradini) [reply CMakeLists.txt:164]: Can you check PR https://github.com/OlivierLDff/asio.cmake/pull/35 and add a comment there.
Thanks 🙏

**Pull request closed without merging**